my-sms/db/migrate/20210208041633_create_titles.rb

```diff
@@ -0,0 +1,70 @@
+class CreateTitles < ActiveRecord::Migration
+  def up
+    available_titles = %w[Mr Mrs Miss Ms Master]
+
+    # do not perform migration if any unexpected titles found in Students
+    # may be more efficient to do via SQL query?
+    Student.find_in_batches do |batch|
```

Comment:
> Would this work instead? Student.where("title NOT IN (?)", available_titles).

That's better. I was trying to avoid SQL queries. Should probably brush up on my SQL.

> Ideally, we would not be doing this inside migrations or at least we would extract into a service class (Migration::ReportStudentTitles) so we can test it properly. And something like Migration::UpdateStudentTitles for what you do after.

As in extract those out into their own classes then call them from the migrations?

> Also, when you raise an error, unless you catch it elsewhere, the migration will stop. Not ideal here since creating titles table is independent of bad data. So you created a dependency which could be avoided.

This was by design. Migrating the old title column to the new one is dependent on a set of titles, established before or during the migration. The migration does not run if this dependency has not been met as the decision of what to do with unexpected titles has not been made.
This is just me trying to assume the hypothetical business context around the SMS and not be destructive with the data, as these issues weren't addressed in the requirements. I'm happy to implement in whatever way is best.

> There is quite a lot happening here which could have been done in separate steps:
1- Identifying/reporting on non-applicable or invalid student titles (independent process, no blockers);
2- Creating titles table and adding defaults (independent process, no blockers - decide what the defaults are);
3- Migrate student titles (dependent on titles table - here we would know what to do about invalid titles found in step one);

Understood. To be clear, these should be separate migrations, and the identifying/reporting is its own process that can be run as a check before migration?

Reply:
@josh-cooper as you said, it is a design decision. I'm all for removing blockers instead of creating them. The current scenario is that your migration will fail if any record managed to go through with a title which is not in the list of `available_titles`. So my question is who has decided what `available_titles` should be and based on what requirement?

There were no validations to prevent invalid titles before and no data source for valid titles. That's why the first step would be running something like:
```
Student.select("DISTINCT(title)")
```
This would give you an idea of what "users" have been filling in and how many discrepancies there are between what the requirement defines as a valid title.

You would have to decide how your migration should handle these discrepancies, e.g: for titles that are not accepted as valid, leave student title blank or set some default value.

Reply:
Thanks I've refactored the migration into multiple steps and also incorporated some of Jordan's comments in 453cd248af4c2a2ecca026d7bcae849771f6fa83
Titles are now added if they are unrecognised in the second part of the migration.